packages/api/internal/orchestrator/client.go

```diff
@@ -177,6 +69,62 @@ func (o *Orchestrator) GetClient(ctx context.Context, clusterID uuid.UUID, nodeI
 		return nil, nil, fmt.Errorf("node '%s' not found in cluster '%s'", nodeID, clusterID)
 	}
 
-	client, ctx := n.getClient(ctx)
+	client, ctx := n.GetClient(ctx)
 	return client, ctx, nil
 }
+
+func (o *Orchestrator) listNomadNodes(ctx context.Context) ([]nodes.NomadServiceDiscovery, error) {
+	_, listSpan := o.tracer.Start(ctx, "list-nomad-nodes")
+	defer listSpan.End()
+
+	options := &nomadapi.QueryOptions{
+		// TODO: Use variable for node pool name ("default")
+		Filter: "Status == \"ready\" and NodePool == \"default\"",
+	}
+	nomadNodes, _, err := o.nomadClient.Nodes().List(options.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]nodes.NomadServiceDiscovery, 0, len(nomadNodes))
+	for _, n := range nomadNodes {
+		result = append(result, nodes.NomadServiceDiscovery{
+			NomadNodeShortID:    n.ID[:consts.NodeIDLength],
+			OrchestratorAddress: fmt.Sprintf("%s:%s", n.Address, consts.OrchestratorPort),
+			IPAddress:           n.Address,
+		})
+	}
+
+	return result, nil
+}
+
+func (o *Orchestrator) GetNode(clusterID uuid.UUID, nodeID string) *nodes.Node {
+	scopedKey := o.scopedNodeID(clusterID, nodeID)
+	n, _ := o.nodes.Get(scopedKey)
+	return n
+}
+
+func (o *Orchestrator) GetClusterNodes(clusterID uuid.UUID) []*nodes.Node {
+	clusterNodes := make([]*nodes.Node, 0)
+	for _, n := range o.nodes.Items() {
+		if n.ClusterID == clusterID {
+			clusterNodes = append(clusterNodes, n)
+		}
+	}
+
+	return clusterNodes
+}
```

Comment:
If we dont want to do this for every node allocation call, we can store nodes in [clusterID][nodeID] map to speed up?

Reply:
Agree